### PR TITLE
feat(internal): add sync tee util

### PR DESCRIPTION
**Pull request by mhofman**
refs: #6943

## Description

This add a util to perform synchronous tee-ing of streams to the internal package. This really should be an `@endo/stream` tool, but I really need this now.

### Security Considerations

None

### Scaling Considerations

Avoid unbounded memory usage by synchronously consuming.

### Documentation Considerations

Should move to endo

### Testing Considerations

Added unit tests

### CODE REVIEW

File: packages/internal/test/test-utils.js [hunk 1]
```diff
@@ -161,3 +162,99 @@ test('forever', async t => {
...
+  const output1 = /** @type {number[]} */ ([]);
+  const output2 = /** @type {number[]} */ ([]);
+  const [reader1, reader2] = syncTee(source, 2);
+  await Promise.all([
+    consumeStreamInto(reader1, output1, 2),
+    consumeStreamInto(reader2, output2),
+  ]);
+
+  t.deepEqual(output1, sourceData.slice(0, 2));
+  t.deepEqual(output2, sourceData.slice(0, 2));
```
Comment (warner) [inline packages/internal/test/test-utils.js:223]: Should this also test that the original stream still has one item left? Or is that not a property we care about?

File: packages/internal/src/utils.js [hunk 2]
```diff
@@ -348,3 +349,101 @@ export const allValues = async obj => {
...
+/**
+ * A tee implementation where all readers are synchronized with each other.
+ * They all consume the source stream in lockstep, and any one returning or
+ * throwing early will affect the others.
+ *
+ * @template [T=unknown]
+ * @param {AsyncIterator<T, void, void>} stream
+ * @param {number} quantity
+ */
+export const syncTee = (stream, quantity) => {
```
Comment (warner) [inline packages/internal/src/utils.js]: Given that this is very much about *async* iterators, maybe name it `synchronizedTee` or `lockstepTee`?

Also, maybe `numReaders` or `readerCount` instead of `quantity`, to make the docstring a bit more instructive.

File: packages/internal/src/utils.js [hunk 3]
```diff
@@ -348,3 +349,101 @@ export const allValues = async obj => {
...
+  /** @type {import('@endo/stream').AsyncQueue<QueuePayload>[]} */
+  const queues = [];
+
+  /** @returns {Promise<void>} */
+  const pullNext = async () => {
+    const requests = await Promise.allSettled(queues.map(queue => queue.get()));
+    const rejections = [];
+    /** @type {Array<(value: PromiseLike<IteratorResult<T>>) => void>} */
+    const resolvers = [];
+    let done = false;
```
Comment (warner) [inline packages/internal/src/utils.js:376]: By "done" we mean "one of the readers stopped reading early", right? Maybe use `readerDone` or `readerStopped` or something similar? Just trying to capture the fruits of my ignorance while I still have it :)

File: packages/internal/src/utils.js [hunk 4]
```diff
@@ -348,3 +349,101 @@ export const allValues = async obj => {
...
+    }
+    /** @type {Promise<IteratorResult<T>>} */
+    let result;
+    if (doneResult) {
+      result = Promise.resolve(doneResult);
+    } else if (rejections.length) {
+      const error = assert.error(assert.details`Teed stream threw`);
+      assert.note(error, assert.details`Teed rejections: ${rejections}`);
+      result =
+        stream.throw?.(error) ||
```
Comment (warner) [inline packages/internal/src/utils.js]: I don't know the semantics of `stream.throw`, but could it throw an error? And would that cause the chain of promises to break without telling the remaining readers? And, could it return a non-promise? In which case the `result.then` below would fail.

I see that any rejected promise by `stream.return` will still proceed to the notify-readers at the end of this function, so I think we're defensive against that method.

File: packages/internal/test/test-utils.js [hunk 5]
```diff
@@ -161,3 +162,99 @@ test('forever', async t => {
   }
   t.is(count, 4, 'count is expected');
 });
+
+/**
+ * @template T
+ * @param {AsyncIterable<T>} stream
+ * @param {Array<T>} output
+ * @param {number} [breakAfter]
```
Comment (warner) [inline packages/internal/test/test-utils.js]: This is like `maxItems`, right? It's counting units of whatever the iterator is iterating over.

**Review by warner — APPROVED** [review]
I'd rely upon @kriskowal for a proper review, but the fact that this passes the unit tests below gives me confidence that it's working as intended.f

**Comment by warner** [comment]
that test failure looked spurious (some web thing), so I kicked it

File: packages/internal/src/utils.js (hunk 3 above)
Comment (mhofman) [reply packages/internal/src/utils.js:376]: `done` is really the term used in `IteratorResults` which have the shape of `{done: boolean, value: any}`.

File: packages/internal/src/utils.js (hunk 4 above)
Comment (mhofman) [reply packages/internal/src/utils.js]: > I don't know the semantics of `stream.throw`, but could it throw an error?

> And, could it return a non-promise?

`.throw` usually rejects yes, which is why we capture the rejection. Unless it unleashes Zalgo, it should not throw synchronously. This is definitely expecting a well behaved source.

> without telling the remaining readers

We do feed the result (likely a rejection) to all the readers. There is a test for that.

> `stream.return` ...  I think we're defensive against that method.

We're not any more defensive than for `throw`, but we make sure that if `return` does exist and returns a promise, that we end up rejecting instead like a generator source would have. These edge cases are not covered by tests however.

File: packages/internal/test/test-utils.js (hunk 1 above)
Comment (mhofman) [reply packages/internal/test/test-utils.js:223]: Modified the last test to be explicit about this.

**Comment by mhofman** [comment]
> I'd rely upon @kriskowal for a proper review

This is inspired by his work, and I shared pretty much this implementation with him earlier. I'd love a proper review, but this is blocking an upcoming PR.

File: packages/internal/src/utils.js [hunk 6]
```diff
@@ -348,3 +349,104 @@ export const allValues = async obj => {
...
+/**
+ * A tee implementation where all readers are synchronized with each other.
+ * They all consume the source stream in lockstep, and any one returning or
+ * throwing early will affect the others.
+ *
+ * @template [T=unknown]
+ * @param {AsyncIterator<T, void, void>} sourceStream
+ * @param {number} readerCount
+ */
+export const synchronizedTee = (sourceStream, readerCount) => {
```
Comment (kriskowal) [inline packages/internal/src/utils.js:362]: `synchronized` is an unnecessary qualifier. All streams are synchronized by default and can be made resilient to bursts by priming a buffer to each individual consumer. In the context of the `@endo/stream` library, `makePipe` returns a synchronized reader/writer which can be buffered by pre-acking a certain number of iterations, effectively allowing the producer to get exactly that far ahead of the consumer. Imagine:

```js
AsyncIterator.prototype.tee(quantity: length): Array<AsyncIterator>;
AsyncIterator.prototype.buffer(length: number): AsyncIterator;
```

File: packages/internal/src/utils.js [hunk 7]
```diff
@@ -348,3 +349,104 @@ export const allValues = async obj => {
...
+      const error = assert.error(assert.details`Teed stream threw`);
+      assert.note(error, assert.details`Teed rejections: ${rejections}`);
+      result =
+        sourceStream.throw?.(error) ||
+        Promise.resolve(sourceStream.return?.()).then(() =>
+          Promise.reject(error),
+        );
+    } else if (done) {
+      result =
+        sourceStream.return?.() ||
```
Comment (kriskowal) [inline packages/internal/src/utils.js:399]: Suggest `??` over `||` here.

File: packages/internal/src/utils.js [hunk 8]
```diff
@@ -348,3 +349,104 @@ export const allValues = async obj => {
...
+      result = sourceStream.next();
+    }
+    result.then(
+      r => {
+        if (r.done) {
+          doneResult = r;
+        }
+      },
+      () => {
+        doneResult = { done: true, value: undefined };
```
Comment (kriskowal) [inline packages/internal/src/utils.js:411]: Consider hardening the result.

File: packages/internal/src/utils.js [hunk 9]
```diff
@@ -348,3 +349,104 @@ export const allValues = async obj => {
...
+      },
+    );
+    resolvers.forEach(resolve => resolve(result));
+    return pullNext();
+  };
+
+  const readers = Array.from({ length: readerCount }).map(() => {
```
Comment (kriskowal) [inline packages/internal/src/utils.js:418]: TIL `Array.from({length})`! I’ve been reluctantly `new Array(length).fill(undefined)`.

File: packages/internal/src/utils.js (hunk 7 above)
Comment (kriskowal) [reply packages/internal/src/utils.js:399]: Once you have an array of promises that you know have settled, some of which may be rejected, it might be sufficient to lean on `Promise.all` to consolidate them into a rejected promise with an `AggregateError` instead of rolling that by hand.

But, it would suffice here to leave a comment with the reason you did not do that, which appears to be to connect with SES causal tracing.

Comment (kriskowal) [reply packages/internal/src/utils.js:399]: Tracking taming Promise.all, worth referencing too https://github.com/endojs/endo/issues/1568

File: packages/internal/src/utils.js [hunk 10]
```diff
@@ -348,3 +349,104 @@ export const allValues = async obj => {
...
+      },
+      // eslint-disable-next-line no-restricted-globals
+      [Symbol.asyncIterator]() {
+        return reader;
+      },
+    });
+    return reader;
+  });
+
+  void pullNext();
```
Comment (kriskowal) [inline packages/internal/src/utils.js:450]: We need to be extra certain that no rejections fall through to this promise. Ideally we could fall through here to `.catch(reportError)` but that is not present in Node.js yet.

File: packages/internal/test/test-utils.js [hunk 11]
```diff
@@ -161,3 +162,104 @@ test('forever', async t => {
...
   t.is(count, 4, 'count is expected');
 });
+
+/**
+ * @template T
+ * @param {AsyncIterable<T>} stream
+ * @param {Array<T>} output
+ * @param {number} [maxItems]
+ */
+const consumeStreamInto = async (stream, output, maxItems) => {
```
Comment (kriskowal) [inline packages/internal/test/test-utils.js:172]: Sometimes spelled `arrayFromAsync`

File: packages/internal/src/utils.js (hunk 6 above)
Comment (mhofman) [reply packages/internal/src/utils.js:362]: I think there can be non synchronized teeing. In fact I have a similar tee helper in another branch where the source is consumed at the pace of the fastest branch (allowing for unbounded memory growth).

Buffering is not equivalent as by definition that only allows for a certain bounded amount of deviation.

File: packages/internal/src/utils.js (hunk 7 above)
Comment (mhofman) [reply packages/internal/src/utils.js:399]: Promise.all short circuits on rejection, which mean aggregation is not possible. I have in fact a `PromiseAllOrRejections` helper that does not short circuit, but alas it's not standard (yet?)

File: packages/internal/test/test-utils.js (hunk 11 above)
Comment (mhofman) [reply packages/internal/test/test-utils.js:172]: Except for the inversion where the "output" array is passed as argument ;)

File: packages/internal/src/utils.js (hunk 6 above)
Comment (kriskowal) [reply packages/internal/src/utils.js:362]: I’d name these to encourage the use of bounded memory and probably wouldn’t even suggest standardizing an unbounded variant with the exception of terminal collectors like `Array.fromAsync` or `response.json()`.
